File: java-core-plugins/src/main/java/org/unlaxer/tramli/plugins/idempotency/InMemoryIdempotencyRegistry.java
package org.unlaxer.tramli.plugins.idempotency;

import java.util.HashSet;
import java.util.Set;

public final class InMemoryIdempotencyRegistry implements IdempotencyRegistry {
    private final Set<String> seen = new HashSet<>();

    @Override
    public synchronized boolean markIfFirstSeen(String flowId, String commandId) {
        return seen.add(flowId + "::" + commandId);
    }
}

File: ts-plugins/src/idempotency/in-memory-idempotency-registry.ts
import type { IdempotencyRegistry } from './types.js';

export class InMemoryIdempotencyRegistry implements IdempotencyRegistry {
  private readonly seen = new Set<string>();

  markIfFirstSeen(flowId: string, commandId: string): boolean {
    const key = `${flowId}::${commandId}`;
    if (this.seen.has(key)) return false;
    this.seen.add(key);
    return true;
  }
}
